### DISCUSSION TITLE
`Enumerable.Range` to return a custom object with fast value enumerator

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [dotnet/runtime#58342] [Ideas] [answered]
[Enumerable.Range](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.range?view=net-5.0) currently returns `IEnumerable<int>` whose enumerator is [implemented](https://source.dot.net/#System.Linq/System/Linq/Range.cs,31) on reference types and virtual calles.

I suggest, instead, returning a value type, whose enumerator would also be a value type. Thanks to inlining it may almost reach the naive implementation of `for` loop, and hence, it would be much easier to use `Enumerable.Range` in `foreach` loops.

Thoughts?

**Comment by stephentoub** [comment, marked as answer]
Return type is part of a method's signature.  Changing this would be a binary breaking change.
https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/breaking-change-rules.md#signatures

**Reply by WhiteBlackGoose** [reply]
I see. But then wouldn't it make sense to add `Enumerable.ValueRange` for the same purpose but fast one?

I managed to make value enumerator as fast as `for`-loop (under some circumstances), whereas the current Enumerable.Range works 15 times slower.

**Reply by stephentoub** [reply]
I don't personally think it adds enough value.  Just write a for loop.  LINQ is about composition with other enumerables and combinators.

**Reply by WhiteBlackGoose** [reply]
Hm. Since it will only improve the experience with `foreach`, this indeed may not be very useful, thanks for replies!
